hamilton/function_modifiers/macros.py

```diff
@@ -859,6 +859,185 @@ def optional_config(self) -> Dict[str, Any]:
 
 
 class post_pipe(base.SingleNodeNodeTransformer):
+    """Running a series of transformation on the output of the function.
+
+    The decorated function declares the dependency, the body of the function gets executed, and then
+    we run a series of transformations on the result of the function specified by `post_pipe`.
+
+    If we have nodes A --> B --> C in the DAG and decorate B with `post_pipe` like
+
+    @post_pipe(step(B1),step(B2))
+    def B(...):
+        return ...
+
+    we obtain the new DAG
+
+    A --> B_raw --> B1 --> B2 --> B --> C,
+
+    where we can think of the B_raw --> B1 --> B2 --> B as a "pipe" that takes the raw output of B, applies to it
+    B1, takes the output of B1 applies to it B2 and then gets renamed to B to re-connect to the rest of the DAG.
+
+    While it is generally reasonable to contain these constructs within a node's function,
+    you should consider `post_pipe` for similar reasons as `pipe`, namely, for any of the following reasons:
+
+    1.  You want the transformations to display as nodes in the DAG, with the possibility of storing or visualizing
+    the result
+    2. You want to pull in functions from an external repository, and build the DAG a little more procedurally
+    3. You want to use the same function multiple times, but with different parameters -- while `@does`/`@parameterize` can
+    do this, this presents an easier way to do this, especially in a chain.
+
+    To demonstrate the rules for chaining nodes, we'll be using the following example. This is
+    using primitives to demonstrate, but as hamilton is just functions of any python objects, this works perfectly with
+    dataframes, series, etc...
+
+
+    .. code-block:: python
+        :name: Simple @post_pipe example
+
+        from hamilton.function_modifiers import step, post_pipe, value, source
+
+
+        def _add_one(x: int) -> int:
+            return x + 1
+
+
+        def _sum(x: int, y: int) -> int:
+            return x + y
+
+
+        def _multiply(x: int, y: int, z: int = 10) -> int:
+            return x * y * z
+
+
+        @post_pipe(
+            step(_add_one),
+            step(_multiply, y=2),
+            step(_sum, y=value(3)),
+            step(_multiply, y=source("upstream_node_to_multiply")),
+        )
+        def run_function(upstream_int: int) -> int:
+            return upstream_int
+
+    .. code-block:: python
```

Comment:
Hmm now that I think about it the nested stuff is pretty confusing...

Reply:
I am leaning towards `@transform_input` and `@transform_output`, because they clarify what it does. Reading pipe sometimes can be confusing.. Maybe others have some strong opinions?

Also, I don't like `function_raw`, but I don't have a better name for it...

Reply:
Re docs:I took the nested part out because the procedural reads better to me and I don't think we need both.

Reply:
So I prefer those too, but talked internally with @skrawcz and @zilto and they both prefer `@pipe_input` and `@pipe_output`. While it's a little less clear exactly what's happening, it matches some common patterns (SKLearn, pandas, polars, etc...), and thus is a a little easier for the people using it to adopt. I'm inclined to go with that -- it's fairly clear as to what's happening too IMO.